### DISCUSSION TITLE
Flags and interrupts confusion (HELP!)

### DISCUSSION THREAD

**Discussion by richonguzman** [jgromes/RadioLib#1066] [General]
Hi

I'm writing LoRa APRS firmware using Radiolib.h for sx126x modules

I've been experimenting lots to only use Radiolib and all boards (and ditch old LoRa.h) but something happened on testing few SX1262 modules:

I need to Tx a packet and then get into Rx packets, nothing crazy there... but after Tx I get always same packet in Rx. The "setFlag" process/function always gets me confused as there is not info about if the setFlag is processed or not and don't know how to process the interrupt/blocking of the Tx and Rx 

any help?

**Comment by jgromes** [comment]
I'm not entirely sure what exactly you're trying to do, but I would suggest to take a look at the [SX126x ping pong example](https://github.com/jgromes/RadioLib/blob/master/examples/SX126x/SX126x_PingPong/SX126x_PingPong.ino) which shows handling of both transmit and receive interrupt. As a side note there's also a [LoRa APRS example](https://github.com/jgromes/RadioLib/blob/master/examples/APRS/APRS_Position_LoRa/APRS_Position_LoRa.ino), though that one only transmits.

**Comment by richonguzman** [comment]
hi

i want to understand the correct procedure to use radiolib and setflag

this is my repository [https://github.com/richonguzman/LoRa_APRS_Tracker](url) which sends and receives data over LoRa into and from APRS.


the problem I'm experimenting is that when I Tx I also Rx the same packet

**Comment by Szetya** [comment]
I'll try to tell you.
The radio IRQ PIN indicates not only reception but also the end of transmission. Therefore the send will also flip the FLAG. If the flag after sending becomes true then the FIFO will read the packet you sent.
Disable or do not allow the receivedFlag to change to true during the send.
 ....
volatile bool enableInterrupt = true;
volatile bool receivedFlag = false;
....
 
void setFlag(void) {
  // check if the interrupt is enabled
  if(!enableInterrupt) {
    return;
  }
  // we sent aor received packet, set the flag
  receivedFlag = true;
  }

void setup() {
...
radio.setPacketReceivedAction(setFlag);
...
}

viod send() {
...
enableInterrupt = false;  
radio.transmit("Anything");
enableInterrupt = true;
...
}

void receive() {
if (receivedFlag) {
receivedFlag = false;
radio.readData();
...
}
}
@jgromes Confirm me, am I saying it right?!

**Reply by jgromes** [reply]
@richonguzman did you take a look at the SX126x I linked in my previous post?

@Szetya you are partially correct in that on SX126x, there is only one interrupt pin that does all the signalling. Therefore, the user must be the one to determine whether the interrupt was caused by transmission or reception. This can be done either by keeping an internal flag (like the SX126x ping pong example does), or you can call `SX126x::getIrqStatus` to check what caused the interrupt signal.

**Comment by richonguzman** [comment]
@jgromes yes I did ;) 

So my goal is to have the LoRa Station always listening and only on certain occasions Tx

so , if I understand correctly the interrupt goes into blocking Rx when Tx and then leaving the interrupt?

**Reply by jgromes** [reply]
> so , if I understand correctly the interrupt goes into blocking Rx when Tx and then leaving the interrupt?

I do not understand that what you are trying to say. Your logic should look like this

1. Start reception in interrupt mode
2. When it's time to transmit, set some variable that tells you you are transmitting.
3. Start transmitting
4. Each time `setFlag` is triggered, check that variable. If it set `setFlag` was triggered by the transmission, and you have to start reception again. If it is not set, you have received data.

**Comment by richonguzman** [comment]
so Setflag is always triggered by Tx and Rx .... any time Tx comes after Rx ... setflag .... any time Rx after Tx setflag....

but if it was just Rx... to Rx again won't set SetFlag again. right?

**Reply by jgromes** [reply]
What? I have no idea what you're trying to say. I would suggest reading up on how interrupts and interrupt service routines work.

**Reply by HeadBoffin** [reply]
And state machines - as you need a variable to know if you are in an Rx state or doing a Tx to be able to figure what the flags & interrupts mean.

**Comment by Szetya** [comment]
Maybe it will help you understand. I've simplified it a lot for clarity but the essentials are there.
Perhaps nothing is missing.

`SX1262 radio = new Module(RADIO_CS_PIN, RADIO_DIO1_PIN, RADIO_RST_PIN, RADIO_BUSY_PIN);     // RADIO_DIO1_PIN is the interrupt PIN
volatile bool enableInterrupt = true;
volatile bool receivedFlag = false;

// Let's assume (I know) you're using an ESP32, so I'm omitting the #ifdef directive for the ICACHE part
// The ICACHE_RAM_ATTR macro places the interrupt handler functions into the correct memory area
ICACHE_RAM_ATTR void setFlag(void) {
  if(!enableInterrupt) {
    return;
  }
  receivedFlag = true;
}

void setup() {
    // Initializes SPI communication with the appropriate pins
    SPI.begin(RADIO_SCLK_PIN, RADIO_MISO_PIN, RADIO_MOSI_PIN);
    // Initializes the radio module
    radio.begin();      // I omit all other things (state=, freq, SF, BW...) for the sake of code length
    // Sets up the TCXO
    radio.setTCXO(DIO3_TCXO_REF, 5000);
    // Sets up CRC
    radio.setCRC();
    // Specifies the action to be taken upon receiving a packet
    radio.setPacketReceivedAction(setFlag);     // I chose this because it's compatible with other radios (No need to choose between setDio0Action and setDio1Action.)
    // Starts radio reception
    radio.startReceive();
}

void send(String transmitPacket) {
    // Disables interrupts during data transmission
    enableInterrupt = false;
    // Transmits data via the radio
    radio.transmit(transmitPacket);
    // Finishes transmission
    radio.finishTransmit();
    // Re-enables interrupts
    enableInterrupt = true;
    // Restarts radio reception for further incoming packets
    radio.startReceive();
}

String receive() {
    String receivedPacket;
    // Checks if there is incoming data
    if (receivedFlag) {
        receivedFlag = false;
        // Reads data from the radio
        radio.readData(receivedPacket);     // I omit all other things like "state="
        // Restarts radio reception for further incoming packets
        // if needed, 'listen' mode can be disabled by calling
        // any of the following methods:
        //
        // radio.standby()
        // radio.sleep()
        // radio.transmit();
        // radio.receive();
        // radio.readData();
        // radio.scanChannel();
        radio.startReceive();
        return receivedPacket;
    }
    return receivedPacket;
}
`

**Comment by richonguzman** [comment]
(I read a few pages and examples and come back with this: ) The goal was to be in Rx mode and, only when/during Tx, block que Rx part (to avoid Rx the Tx packet)

`#ifndef LORA_EXPLANATION_H_
#define LORA_EXPLANATION_H_

#include <Arduino.h>
#include <RadioLib.h>
#include "pins_config.h"

extern String packet;

SX1262 radio = new Module(RADIO_CS_PIN, RADIO_DIO1_PIN, RADIO_RST_PIN, RADIO_BUSY_PIN);
bool transmissionFlag = true;
bool enableInterrupt = true;

namespace Radio_Utils {

    void setFlag() {
        if(!enableInterrupt) {
            return;
        }
        transmissionFlag = true;
    }

    void setup() {
        SPI.begin(RADIO_SCLK_PIN, RADIO_MISO_PIN, RADIO_MOSI_PIN);
        float freq = 433.775;
        int state = radio.begin(freq);
        if (state == RADIOLIB_ERR_NONE) {
            Serial.println("Initializing SX126X");
        } else {
            Serial.println("Starting LoRa failed!");
            while (true);
        }
        radio.setDio1Action(setFlag);
        radio.setSpreadingFactor(12);
        float signalBandwidth = 125.0;
        radio.setBandwidth(signalBandwidth);
        radio.setCodingRate(5);
        radio.setCRC(true);
        radio.setOutputPower(22);
        radio.setCurrentLimit(140);
        radio.setRxBoostedGainMode(true);
        radio.startReceive();
    }

    void txPacket(const String &newPacket) {
        Serial.print("LoRa Tx ---> "); Serial.println(newPacket);
        enableInterrupt = false;
        int state = radio.transmit(newPacket.c_str());
        if (state == RADIOLIB_ERR_NONE) {
            //Serial.println(F("success!"));
        } else {
            Serial.print(F("failed, code "));             Serial.println(state);
        }
        enableInterrupt = true;
        radio.startReceive();
    }

    void rxPacket() {
        String packet = "";
        if (transmissionFlag) {
            transmissionFlag = false;
            int state = radio.readData(packet);
            if (state == RADIOLIB_ERR_NONE) {
                if(!packet.isEmpty()) {
                    Serial.print("LoRa Rx ---> "); Serial.println(packet);
                }
            } else {
                Serial.print(F("failed, code ")); Serial.println(state);
            }
        }
    }

}

#endif`


does it look right to you?

**Reply by Crsarmv7l** [reply]
What all of these nice people are getting at is the transceiver can only be in either tx OR rx, not both at the same time. They are half duplex. The flag allows you to switch between rx and tx. But you need to come up with the logic.

**Reply by richonguzman** [reply]
I finally figure it out with all the help comments in here...

I had the Tx interrupt not in the code and this made the Rx still get the recently Tx packet.

so now the Rx is disabled while Tx and then it gets back on. Working awesome!
